### DISCUSSION TITLE
How to download Ivy dependencies "explicitly" taking the Scala version into account

### DISCUSSION THREAD

**Discussion by hmf** [com-lihaoyi/mill#2045] [Q&A] [answered]
For 

```scala
  override def scalaVersion = T{ "3.1.3" }
```

I have the following code:

```scala
  def mDocLibs = 
      T{
        import coursier._
        val files = Fetch().addDependencies(dep"org.scalameta::mdoc:2.3.3").run()
        val pathRefs = files.map(f => PathRef(os.Path(f)))
        Agg(pathRefs : _*)
      }
```
In the project's module I also have:

```scala
  override def ivyDeps = Agg(
      ivy"org.tensorflow:tensorflow-core-api:${tensorflowVersion}",
      ivy"org.tensorflow:tensorflow-core-api:${tensorflowVersion};classifier=${javaCPPPlatform()}",
      ivy"org.scalameta::mdoc:2.3.3"
  )
```
If I compare the output (just focusing on MDoc): 

```scala
    val rp = mDocLibs().map(_.path)
    val cp = runClasspath().map(_.path)

    val s1 = Set(cp.toList:_*)
    val s2 = Set(rp.toList:_*)
    val s = s1 -- s2
```

I see that for the explicit download I have:
```
home/usercache/coursier/v1/https/repo1.maven.org/maven2/org/scalameta/mdoc_2.13/2.3.3/mdoc_2.13-2.3.3.jar
```
and its dependencies. But for the standard `ivyDeps` I get: 

```
/home/user/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scalameta/mdoc_3/2.3.3/mdoc_3-2.3.3.jar
```
and its dependencies. So what Mill API can I use to generate the `ivyDeps` equivalent taking the Scala version into account?

TIA

**Comment by lefou** [comment]
`ivyDeps` with a double-collon as the first separator will use the same Scala version that is configured for the module. You can use `ivy"org.scalameta::mdoc:2.3.3".withDottyCompat(scalaVersion())` to use a compatible Scala 2.13 dependency.

**Comment by hmf** [comment, marked as answer]
@lefou  thanks for the feedback. I am assuming that you are referring to the use of `override def ivyDeps`. However, I need to download these dependencies separately so that I can use them like this:

```scala
  override def mdoc : T[PathRef] = T {
  
    val rp = mDocLibs().map(_.path)
    val cp = runClasspath().map(_.path)

    val dir = T.dest.toIO.getAbsolutePath
    val dirParams = mdocSources().map(pr => Seq(
        "--classpath", toArgument(cp),
        "--in", pr.path.toIO.getAbsolutePath, 
        "--out",  dir)
      ).iterator.flatten.toSeq
  
    Jvm.runLocal("mdoc.Main", rp, dirParams)
  
    PathRef(T.dest)
  }
``` 

Note how the `cp` uses `ivyDeps` but `rp` does not. `rp` should just contain the dependencies of one specific library - MDoc.  The use of Coursier's  `Fetch().addDependencies` does not take the Scala version into account. I would like to use Mill to identify, download and report on the dependencies as it does using `ivyDeps` to generate the `runClasspath`.

Hope this makes sense. 

Just as an aside. I can use the `ivyDeps` with the double colon and do not require to add `.withDottyCompat(scalaVersion())`. It works perfectly. 8-)

**Reply by hmf** [reply]
I am one step closer. If I use:

```scala
        val files = Fetch().addDependencies(dep"org.scalameta:mdoc_3:2.3.3").run()
```

So the question is, how does Mill to handle this Scala 2 versus 3 issue? Is it some Coursier parameter? I can see Coursier also handles the `::` "interop" format but still have not a way to use it correctly.

**Reply by lefou** [reply]
In the end, it's just some convenience to add a `_2.13` or `_3` suffix to the artifact name.

If you look into how `ivyDeps` are resolved (hint: `resolvedIvyDeps`), you'll find `CoursierModule.resolveDeps`, which you also can use directly to implement `mDocLibs`.

**Reply by hmf** [reply]
Exactly what I was looking for. Having difficulty tracking this down via the Git's search. Worked like a charm.  Thank you.
